### PR TITLE
core::str::{chars_uppercase,chars_lowercase} iterators

### PR DESCRIPTION
**Pull request by markokr**
They are based on new UnicodeConverter + UnicodeIterator
internal API that supports context-sensitivity and char expansion.

API change proposal: https://github.com/rust-lang/libs-team/issues/58

### CODE REVIEW

**Comment by thomcc** [comment]
This is a new API, so please file an API change proposal (see the links in the bot post). That said, I'm very much in favor of this, and a quick skim (will need more thorough review once it gets API sign off) looks fine.

@rustbot label +T-libs-api -T-libs

**Comment by markokr** [comment]
Thanks for quick feedback!

Created API proposal here: https://github.com/rust-lang/libs-team/issues/58

File: library/core/src/str/iter.rs [hunk 1]
```diff
@@ -1497,3 +1498,346 @@ macro_rules! escape_types_impls {
 }
 
 escape_types_impls!(EscapeDebug, EscapeDefault, EscapeUnicode);
+
+/// Internal trait for Unicode conversions that potentially
+/// need context and can expand one char to several.
+///
+/// Default implementation is pass-through, no conversion is done,
+/// with `is_simple = is_ascii`.
+trait UnicodeConverter {
```
Comment (thomcc) [inline library/core/src/str/iter.rs:1507]: I think this shouldn't use a trait. I think this ends up complicating things quite a bit and makes the implementation overly generic.

File: library/core/src/str/iter.rs [hunk 2]
```diff
@@ -1497,3 +1498,346 @@ macro_rules! escape_types_impls {
...
+/// to cache expanded chars.
+#[derive(Clone)]
+struct UnicodeIterator<'a, C>
+where
+    C: UnicodeConverter,
+{
+    // data source
+    iter: CharIndices<'a>,
+    // buffer for .next()
+    fwd: [Option<char>; 2],
```
Comment (thomcc) [inline library/core/src/str/iter.rs:1593]: nit: forward/backward instead of `fwd/bwd`.

File: library/alloc/src/lib.rs [hunk 3]
```diff
@@ -144,6 +144,7 @@
 #![feature(unicode_internals)]
 #![feature(unsize)]
 #![feature(std_internals)]
+#![feature(unicode_converter)]
```
Comment (thomcc) [inline library/alloc/src/lib.rs:147]: This feature name is too generic, perhaps something more like `#![feature(str_chars_casemapped)]`?

**Review by thomcc — CHANGES_REQUESTED** [review]
Hmm, the implementation here is more generic than it needs to be, and more complex as a result. You should not do this via a trait IMO -- The way you've decoupled it here also requires more bounds-checks which shouldn't be needed, and avoiding them in this interface would require otherwise-unnecessary unsafe code (to be clear: I'm not asking for you to add unsafe to this).

I think `chars_lowercase` should be a bit more like what you'd get by manually performing `s.chars().flat_map(|c| c.to_lowercase())`. Uppercase is the complex one, but should be handled directly, closer to the implementation in liballoc.

Let me know if you need an example of what I mean.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@markokr what is the status of this PR?

**Comment by JohnCSimon** [comment]
@markokr 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**